File: Leetcode 0051 N-Queens [hard] [array]/N-Queens [dfs].js
/**
 * @param {number} n
 * @return {string[][]}
 */
var solveNQueens = function(n) {
    let result = [];
    if (n <= 0) {
        return result;
    }
    
    let queensCol = [];
    dfs(0, result, queensCol, n);
    return result;
};

let dfs = (curRow, result, queensCol, size) => {
    if (curRow === size) {
        result.push(draw(queensCol));
        return;
    }
    
    for (let i = 0; i < size; i++) {
        if (isValid(queensCol, curRow, i)) {
            queensCol.push(i);
            dfs(curRow + 1, result, queensCol, size);
            queensCol.pop();
        }
    }
}

let isValid = (queensCol, row, col) => {
    for (let i = 0; i < queensCol.length; i++) {
        if (queensCol[i] === col) {
            return false;
        } else if (row + col === i + queensCol[i]) {
            return false;
        } else if (row - col === i - queensCol[i]) {
            return false;
        }
    }
    return true;
}

let draw = (queensCol) => {
    let chessboard = [];
    for (let i = 0; i < queensCol.length; i++) {
        let sb = "";
        for (let j = 0; j < queensCol.length; j++) {
            sb = sb + (j === queensCol[i] ? 'Q' : '.');
        }
        chessboard.push(sb);
    }
    return chessboard;
}

File: Leetcode 0051 N-Queens [hard] [array]/N-Queens [dfs].java
class Solution {
    public List<List<String>> solveNQueens(int n) {
        List<List<String>> result = new ArrayList<>();
        if (n <= 0) {
            return result;
        }
        
        List<Integer> queensCol = new ArrayList<>();
        dfs(0, queensCol, result, n);
        return result;
    }
    
    private void dfs(int curRow, List<Integer> queensCol, List<List<String>> result, int size) {
        if (curRow == size) {
            result.add(draw(queensCol));
            return;
        }
        
        for (int i = 0; i < size; i++) {
            if (isValid(queensCol, curRow, i)) {
                queensCol.add(i);
                dfs(curRow + 1, queensCol, result, size);
                queensCol.remove(queensCol.size() - 1);
            }
        }
    }

    /**
    判斷是否current coordinate是否合法:
    利用對角線相加, 與相減判斷兩對角線, 及是否在同行已有皇后, 本身遞歸過程已除去同列的可能
    */
    private boolean isValid(List<Integer> queensCol, int row, int col) {
        for (int i = 0; i < queensCol.size(); i++) {
            if (col == queensCol.get(i)) {
                return false;
            } else if (row + col == i + queensCol.get(i)) {
                return false;
            } else if (row - col == i - queensCol.get(i)) {
                return false;
            }
        }
        return true;
    }

    private List<String> draw(List<Integer> queensCol) {
        List<String> chessboard = new ArrayList<>();
        for (int i = 0; i < queensCol.size(); i++) {
            StringBuilder sb = new StringBuilder();
            for (int j = 0; j < queensCol.size(); j++) {
                sb.append(j == queensCol.get(i) ? 'Q' : '.');
            }
            chessboard.add(sb.toString());
        }
        return chessboard;
    }
}